### DISCUSSION TITLE
How can I update the rendering in the execute method?

### DISCUSSION THREAD

**Discussion by deeprobin** [apache/airflow#22108] [Q&A]
I want to fetch the data of a Stored Procedure and fill it into a template field.

Thats my current bad code:
```py
class StoredProcedureOperator(BaseOperator):

    template_fields = ('call_sql', 'fetch_procedure_sql', 'procedure_sql')
    template_ext = ('.sql', '.sql', '.sql')

    def __init__(
            self,
            procedure_name,
            mssql_conn_id='connid', 
            schema='dbo',
            database="Database",
            **kwargs) -> None:
        super().__init__(**kwargs)
        self.mssql_conn_id = mssql_conn_id
        self.database = database

        self.call_sql = f'EXECUTE [{schema}].[{procedure_name}]'
        self.fetch_procedure_sql = f"SELECT [definition] FROM [sys].[sql_modules] WHERE [object_id] = OBJECT_ID('[{schema}].[{procedure_name}]')"
        self.procedure_sql = ""
###
        self.hook = MsSqlHook(mssql_conn_id=self.mssql_conn_id,
                         schema=self.database)
        con = self.hook.get_conn()
        cursor = con.cursor()
        self.log.info('Fetching procedure data %s', self.fetch_procedure_sql)
        cursor.execute(self.fetch_procedure_sql)
        row = cursor.fetchone()

        if not row:
            self.procedure_sql = "-- Procedure not found"
            self.log.info('Procedure not found')
        else:
            self.procedure_sql = row[0]
            self.log.info('%s', self.procedure_sql)

        if self.procedure_sql is None:
            self.procedure_sql = "-- Cannot fetch procedure data"
            self.log.info('Cannot fetch procedure data')
###


    def execute(self, context):
        con = self.hook.get_conn()
        cursor = con.cursor()

        cursor.execute(self.call_sql)
        
        cursor.close()
        con.close()
```

The current procedure is that the code drawn in "###" is executed in the __init__ method.

This is not supposed to be this way, but it is the only way I have found where the rendering view updates.

Do you have any better suggestions?
Can I somehow explicitly tell the Airflow instance in the execute function to re-render the Jinja templates?

**Comment by potiuk** [comment]
Deleted previous answer. You do not need JINJA in this case. Simply move your fetch_procedure_call assignment to execute as well:

In init:
```
       self.database = database
       self.schema = schema
       self,database = database
       self.procedure_name = procedure_name
```  

In execute:

```
        fetch_procedure_sql = f"SELECT [definition] FROM [sys].[sql_modules] WHERE [object_id] = OBJECT_ID('[{self.schema}].[{self.procedure_name}]')"
        self.procedure_sql = ""
        self.call_sql = f'EXECUTE [{schema}].[{procedure_name}]'
...
```

No need to use template_fields in this case nor Jinja.

**Reply by deeprobin** [reply]
However, I would like to display the correspondingly generated SQL in the "Rendered Template", which does not work in execute.

### `procedure_sql` in `__init__`
![grafik](https://user-images.githubusercontent.com/10373452/157454186-f5df5074-6ab5-4014-8f63-a217b703c35e.png)

### `procedure_sql` in `execute`
![grafik](https://user-images.githubusercontent.com/10373452/157454979-34b62f4d-9b38-40c4-b2a1-693615d8ed7a.png)

**Reply by potiuk** [reply]
You can't use OTHER constructor arguments in templated one. You can only use those values: https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html

**Reply by deeprobin** [reply]
But why (imo a misconception)? Is there no way to manually re-render the output in execute?

**Reply by potiuk** [reply]
Rendering s done BEFORE execute. That's the whole point. Rendering results are INPUT to execute not OUTPUT.

**Reply by potiuk** [reply]
You probably COULD use "{{ task.&lt;smth&gt; }}" in your template. This is a Task object - maybe it has all that you need. Try it and see what it has.
